### DISCUSSION TITLE
Best way to model and update dynamic obstacles?

### DISCUSSION THREAD

**Discussion by sea-bass** [stack-of-tasks/pinocchio#2379] [Q&A] [answered]
Before I go down the wrong path, I wanted to ask here.

I am trying to use Pinocchio to add "scene obstacles" to a robot model. If the obstacles are static, it's very easy to add geometry objects to the model and keep track of collisions, etc. But what to do about dynamic objects?

Some options I've ran by:

1. Create a floating joint between the universe and the object. Works great for a small number of obstacles, but it blows up the dimension of the state/Jacobians, and the geometry itself cannot change at runtime.
2. Manually remove and re-add geometry objects every time they change. Highly inefficient and just seems wrong.
3. Custom logic that manages all the transforms and performs collision checks with lower-level APIs. Flexible, but would require significant copy paste of code already in Pinocchio.
4. Manually modify the `Data` and `GeometryData` of the models, and then call `updateGeometryPlacements()`. This seems promising, but I'm unsure if this is a recommended workflow.

One special consideration is the ability to have a single Octree in the world, but be able to update its points. I'm sure this must be a path that these tools are designed for...

Is there some recommended approach to follow out of all these? Or something else I have missed in my search? Any help is much appreciated.

---

EDIT: Playing with the Python API, I found you could just... change positions on the fly?

```python
cid = collision_model.getGeometryId("obstacle_0")
collision_model.geometryObjects[cid].placement.translation[2] += 0.05
```

And geometries?

```python
collision_model.geometryObjects[cid].geometry = hppfcl.Sphere(1.5)
```

Is this valid usage? Seemed to work when I tested.

**Comment by jcarpent** [comment, marked as answer]
You can, of course, update the geometry placement as you want. This quantity is considered when you ask for a 0th-order forward kinematics (e.g., `updateGeometryPlacements`, that will update the `geometry_data.oMg` vector. 

You can also update the geometry parameters (e.g., `collision_model.geometryObjects[cid].geometry.radius = 1.5`), but you cannot change the type. Otherwise, you need to recreate the associated data structure (that stores some functors for fast evaluation of collision checking).
If you use the broadphase, when changing the dimensions of a given geometry, you will need to recompute the local AABB encompassing volume, using `collision_model.geometryObjects[cid].geometry.computeLocalAABB()`.
